enterprise/server/remote_execution/containers/firecracker/firecracker.go

```diff
@@ -341,36 +283,168 @@ func NewContainer(ctx context.Context, opts ContainerOpts) (*firecrackerContaine
 	// /tmp/firecracker/217d4de0-4b28-401b-891b-18e087718ad1/root/run/fc.sock
 	// everything after "/tmp" is 65 characters, so 38 are left for the
 	// jailerRoot.
-	jailerRoot := "/tmp/"
-	if err := disk.EnsureDirectoryExists(jailerRoot); err != nil {
+	if opts.JailerRoot == "" {
+		opts.JailerRoot = "/tmp"
+	}
+	if len(opts.JailerRoot) > 38 {
+		return nil, status.InvalidArgumentErrorf("JailerRoot must be < 38 characters. Was %q (%d).", opts.JailerRoot, len(opts.JailerRoot))
+	}
+	if err := disk.EnsureDirectoryExists(opts.JailerRoot); err != nil {
 		return nil, err
 	}
 
-	c := &firecrackerContainer{
-		numCPUs:          opts.NumCPUs,
-		memSizeMB:        opts.MemSizeMB,
-		enableNetworking: opts.EnableNetworking,
-		debugMode:        opts.DebugMode,
-
-		jailerRoot:       jailerRoot,
+	c := &FirecrackerContainer{
+		constants: Constants{
+			NumCPUs:          opts.NumCPUs,
+			MemSizeMB:        opts.MemSizeMB,
+			EnableNetworking: opts.EnableNetworking,
+			DebugMode:        opts.DebugMode,
+		},
+		jailerRoot:       opts.JailerRoot,
 		containerImage:   opts.ContainerImage,
 		actionWorkingDir: opts.ActionWorkingDirectory,
+		env:              env,
 	}
 
 	if err := c.newID(); err != nil {
 		return nil, err
 	}
+	if opts.ForceVMIdx != 0 {
+		c.vmIdx = opts.ForceVMIdx
+	}
 	return c, nil
 }
 
+func (c *FirecrackerContainer) SaveSnapshot(ctx context.Context, instanceName string) (string, error) {
+	if err := c.machine.PauseVM(ctx); err != nil {
+		return "", err
+	}
+	memSnapshotPath := filepath.Join(c.getChroot(), fullDiskSnapshotName)
```

Comment:
i think mem and disk naming is swapped here

Reply:
Done.